pkg/stanza/fileconsumer/file_sort.go

```diff
@@ -0,0 +1,206 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package fileconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer"
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+
+	strptime "github.com/observiq/ctimefmt"
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils"
+)
+
+const (
+	SortTypeNumeric      = "numeric"
+	SortTypeTimestamp    = "timestamp"
+	SortTypeAlphabetical = "alphabetical"
+)
+
+type SortRule struct {
+	Regex     string `mapstructure:"regex,omitempty"`
+	SortType  string `mapstructure:"sort_type,omitempty"`
+	Format    string `mapstructure:"format,omitempty"`
+	Location  string `mapstructure:"location,omitempty"`
+	Ascending bool   `mapstructure:"ascending,omitempty"`
+}
+
+func (f SortRule) validate() error {
+	if f.Regex == "" {
+		return fmt.Errorf("regex must be specified for sort")
+	}
+
+	switch f.SortType {
+	case SortTypeNumeric:
+		if f.Format != "" {
+			return fmt.Errorf("format must be specified for timestamp sort")
+		}
+		if f.Location != "" {
+			return fmt.Errorf("location shouldn't be specified for timestamp sort")
+		}
+		return nil
+	case SortTypeAlphabetical:
+		if f.Format != "" {
+			return fmt.Errorf("format must be specified for timestamp sort")
+		}
+		if f.Location != "" {
+			return fmt.Errorf("location shouldn't be specified for timestamp sort")
+		}
+		return nil
+	case SortTypeTimestamp:
+		if f.Format == "" {
+			return fmt.Errorf("format must be specified for timestamp sort")
+		}
+		if f.Location == "" {
+			return fmt.Errorf("location must be specified for timestamp sort")
+		}
+		_, err := strptime.ToNative(f.Format)
+		if err != nil {
+			return fmt.Errorf("error parsing format %s: %w", f.Format, err)
+		}
+
+		_, err = time.LoadLocation(f.Location)
+		if err != nil {
+			return fmt.Errorf("error parsing location %s: %w", f.Location, err)
+		}
```

Comment:
We have several possible "types" of sort rules, but all possible fields are at the same level. This tends to become problematic over time, and this is an early example of it because validating a given type ends up mostly being "don't use fields that only work for other types".

I think there are better ways to do this. Probably define a struct for each type and make `SortRules` a slice of generics.